zircon.core/src/commonMain/kotlin/org/hexworks/zircon/internal/fragment/impl/table/DefaultTable.kt

```diff
@@ -0,0 +1,127 @@
+package org.hexworks.zircon.internal.fragment.impl.table
+
+import org.hexworks.cobalt.databinding.api.extension.toProperty
+import org.hexworks.cobalt.databinding.api.property.Property
+import org.hexworks.cobalt.databinding.api.value.ObservableValue
+import org.hexworks.zircon.api.Components
+import org.hexworks.zircon.api.component.AttachedComponent
+import org.hexworks.zircon.api.component.Component
+import org.hexworks.zircon.api.component.HBox
+import org.hexworks.zircon.api.component.VBox
+import org.hexworks.zircon.api.data.Size
+import org.hexworks.zircon.api.fragment.Table
+import org.hexworks.zircon.api.uievent.MouseEventType
+import org.hexworks.zircon.api.uievent.UIEventPhase
+import org.hexworks.zircon.api.uievent.UIEventResponse
+
+/**
+ * The **internal** default implementation of [Table].
+ */
+class DefaultTable<M: Any>(
+    private val data: List<M>,
+    private val columns: List<TableColumn<M, *, *>>,
+    /**
+     * The height this fragment will use. Keep in mind that the first row will be used as header row.
+     */
+    fragmentHeight: Int,
+    private val rowSpacing: Int = 0,
+    private val colSpacing: Int = 0
+): Table<M> {
+
+    init {
+        require(data.isNotEmpty()) {
+            "A table may not be empty! Please feed it some data to display."
+        }
```

Comment:
Might be nice to lift this requirement. Would that be possible?

Reply:
I thought "sure, why not!", but when I just tried it I realized that there is `selectedElement: Property<M>`. With an empty list we would need to handle that and I don't know how complicated that would become. I think `Property` can not contain nullable values...

Reply:
Hm...maybe the caller would have to provide a "null object pattern" object then? Not really a fan, but might be the only way?

I don't see it being terribly common but this does seem like something callers would expect to work.

Reply:
I agree with @nanodeath . Imagine that you have filter fields above your table and the table itself has an `ObservableList` so when you type something the list is updated accordingly. It might happen that there are no items that correspond to your query and the table will be empty.

Selection is also a different concept. You can have a table that has `n` elements but none of them are selected.

See my comment above for a workaround for this problem that would also alleviate the need for `null`s and null objects.